### DISCUSSION TITLE
Support for embedded platforms, like ESP-IDF?

### DISCUSSION THREAD

**Discussion by snej** [libuv/libuv#4132] [Ideas]
I'm interested in libuv support for Espressif's ESP32 microcontrollers using the ESP-IDF framework. (ESP32 supports other APIs like Arduino, but they're all just wrappers around ESP-IDF.) 

I have no connection to Espressif, but I've worked with their systems in the past, and I think the hardware and software would be a good fit for libuv:

* 32-bit CPU, usually ~200MHz; ISA is either Xtensa (proprietary) or RISC-V; the official toolchain is GCC. Fully supports C++.
* Depending on the exact product, there's at least about 512KB of internal RAM, 8MB or more of (slower) off-chip PSRAM, and 16MB or so of Flash ROM.
* WiFi (2.4GHz) and Bluetooth networking.
* The OS core is a modified FreeRTOS with some SMP capabilities. FreeRTOS is event-loop based.
* Much of the pthreads API is supported, as a wrapper around FreeRTOS.
* Networking API is basically POSIX sockets. (This is new; I haven't kicked the tires yet. They used to use `lwip`.)
* FAT32 or SPIFFS filesystems, e.g. on SD cards.
* mbedTLS.

I don't mean this to be an ad :) But when I say "embedded", some people might still be thinking of 8-bit Arduinos with 2KB RAM. This is much more capable. It's probably better specced than the PCs I was using in the mid-90s! It just doesn't run a "real" OS.

I haven't been able to find signs that anyone has tried this, so I thought I'd ask here. I'd be interested in putting some time into it, as I'm pretty experienced in C/C++, although I am pretty new at using libuv and have not looked at the implementation yet.

**Comment by bnoordhuis** [comment]
I've toyed with the idea of porting libuv to FreeRTOS (also LK a.k.a. LittleKernel) and I'm reasonably sure it can be done. They're both pretty bare-bones so you'll probably end up with a lot of stub functionality inside libuv but that isn't necessarily a blocker for a tier 3 platform.

What's more of an issue is that RT kernels are sufficiently different from most POSIX platforms that either src/unix turns into (worse than it already is) #ifdef soup, or you end up duplicating a lot of code. Both are not great from a maintenance perspective and that's why I ultimately didn't end up pursuing it; not invested enough.

**Reply by snej** [reply]
Yeah, after posting that I took a look at the libuv source and discovered it’s not at all modular nor designed for extensibility. 😬

I’m already writing an OO C++ framework around libuv, so it will probably be a lot easier to add FreeRTOS / ESP-IDF support directly to my framework.

**Reply by bnoordhuis** [reply]
It's modular enough for what it's designed to do, and no more. :-)
